plugins/storage/pyparts/ceph_cluster_checks.py

```diff
@@ -336,6 +336,76 @@ def get_crushmap_mixed_buckets(self):
             issue = issue_types.CephCrushWarning(msg)
             issue_utils.add_issue(issue)
 
+    def _is_bucket_imbalanced(self, buckets, start_bucket_id, failure_domain):
+        """
+        Recursively determine if a tree is balanced at the given failure domain
+        """
+        unbalanced = False
+        weight = -1
+
+        for item in buckets[start_bucket_id]["items"]:
+            if buckets[item["id"]]["type_name"] != failure_domain:
+                unbalanced = self._is_bucket_imbalanced(buckets,
+                                                        item["id"],
+                                                        failure_domain)
+                if unbalanced:
+                    return unbalanced
+            else:
+                if weight == -1:
+                    weight = item["weight"]
+                else:
+                    if weight != item["weight"]:
+                        unbalanced = True
+                        return unbalanced
+
+        return unbalanced
+
+    def check_crushmap_equal_buckets(self):
+        """
+        Report when buckets of the failure domain type in a
+        CRUSH rule referenced tree are unbalanced.
+
+        Uses the trees and failure domains referenced in the
+        CRUSH rules, and checks that all buckets of the failure
+        domain type in the referenced tree are equal.
+        """
+        osd_crush_dump = self.cli.ceph_osd_crush_dump_json_decoded()
+        if not osd_crush_dump:
+            return
+
+        buckets = {}
+        for bucket in osd_crush_dump["buckets"]:
+            bid = bucket["id"]
+            buckets[bid] = bucket
+
+        to_check = []
+
+        for rule in osd_crush_dump['rules']:
+            taken = 0
+            fdomain = 0
+            for i in rule['steps']:
+                if i["op"] == "take":
+                    taken = i["item"]
+                if "type" in i and taken != 0:
+                    fdomain = i["type"]
+                if taken != 0 and fdomain != 0:
+                    to_check.append((rule["rule_id"], taken, fdomain))
+                    taken = fdomain = 0
```

Comment:
i think the above should be outside of the "if" shouldn't it? i.e. you want to reset before/after each step (not just the ones where you find a match).

Reply:
I want to reset only after both `taken` and `fdomain` are populated. It is an admittedly naive implementation. Resetting on each step in a rule would result in no rules to process.

It is ugly because the tree and failure domain are in separate steps, that you can't be guaranteed are directly adjacent.